<h1>Tableaux de données - matrices</h1>

<div class="alert alert-warning">
    <strong>Objectifs</strong>
    
1. manipuler les fonctions de tri de données dans un tableur (notamment le tri sur plusieurs champs successifs)
2. initialiser des tableaux de données à 2 dimensions en python: on parle alors de matrice.
    
</div>

## Fichier xlsx des établissements en Auvergne-Rhônes-Alpes

Le fichier [etabs_rra.xlsx](https://0692717d.moodle.ent.auvergnerhonealpes.fr/mod/resource/view.php?id=4931) est un fichier tableur des établissements scolaires publics et privés du 2e degré (collèges lycées) dans la région.

→ le copier et l'ouvrir dans un tableur (libreoffice calc ou ms office)

Après avoir activer l'AutoFiltre ([ms](https://support.microsoft.com/fr-fr/office/utiliser-le-filtre-automatique-pour-filtrer-des-donn%C3%A9es-7d87d63e-ebd0-424b-8106-e2ab61133d92), [LO](https://help.libreoffice.org/6.2/fr/text/scalc/guide/autofilter.html)) sur les champs de données, répondez aux questions suivantes:

1. Par ordre croissant de nom_etablissement, quel est le 25e établissement de la région?
2. Et par ordre d'UAI (code d'Unité Administrative Immatriculée), quel est le 25e?
3. Retrouvez l'UAI du lycée Jacques Brel.
4. Parmi les lycées, quel est le lycée le plus au sud de la région? (trier les données par type d'établissement (1ere clé) puis latitude (2e clé))
5. Quel est le lycée le plus à l'Est dans l'Ain? trouvez dans quelle ville il se situe.

## Initialisation de matrice

*Mémo sur les listes*:

* On peut accrocher un élément `e` à la fin d'une liste `L` par l'appel `L.append(e)`.
* On obtient la taille de la liste `L` par l'appel de `len(L)`.
* initialisation d'une liste de taille donnée avec un même élément:

In [2]:
L = 5 * [0] # liste de taille 5, que des 0
M = 10 * ["a"] # liste de taille 10, que des lettres "a"
print(L, M)
mot = "bonjour"
N = list(mot) # liste des caractères de mot
print(N)
# méthode alternative et puissante: définition en compréhension
L = [0 for i in range(5)]
M = ["a" for i in range(10)]
N = [e for e in mot]

[0, 0, 0, 0, 0] ['a', 'a', 'a', 'a', 'a', 'a', 'a', 'a', 'a', 'a']
['b', 'o', 'n', 'j', 'o', 'u', 'r']


Une matrice de taille m×n possède m lignes et n colonnes. On la définit comme la *liste de ses m lignes*.

**exemple 1**: initialisation de `M1` matrice de taille 3×5 remplie de 0 en deux temps

In [4]:
l1 = 5 * [0] # une ligne avec 5 zéros
M1 = [l1 for i in range(3)]
# construction directe
# M1 = [ 5*[0] for i in range(3)]

# affichage avec retour à la ligne pour chaque ligne: méthode 1
for i in range(3):
    print(M1[i])
print("_______________")
# méthode 2: afficher les éléments de M1 séparés par "\n"
print(*M1, sep="\n")
print("_______________")
# méthode 3: boucle sur les éléments(lignes) de M1 directement
for e in M1:
    print(e)

[0, 0, 0, 0, 0]
[0, 0, 0, 0, 0]
[0, 0, 0, 0, 0]
_______________
[0, 0, 0, 0, 0]
[0, 0, 0, 0, 0]
[0, 0, 0, 0, 0]
_______________
[0, 0, 0, 0, 0]
[0, 0, 0, 0, 0]
[0, 0, 0, 0, 0]


**exercice 1**: créez une matrice M de taille 4×10 remplie de 9 et une matrice N de taille 6×2 remplie de "" (chaîne de caractères vide). Affichez M et N.

---
## Remplissage de matrice

Cette section vous indique le principe général de double-boucle imbriquée, mais il reste à votre charge de vous adapter à chaque situation.

**exemple 2**: Construisons une matrice M de taille 10×6 dont

- la 1ere ligne contient les entiers j allant de 1 à 6
- la 2e contient les entiers 2j pour j allant de 1 à 6: 2, 4, …, 12
- la 3e contient les entiers 3j pour j allant de 1 à 6: 3, 6, …, 18
etc.
c'est-à-dire que la i-e ligne contient les entiers i×j (j allant de 1 à 6)
La formule de calcul pour la case (i,j) (i-e ligne, j-e colonnes) est `M[i][j] = (i+1)(j+1)` (oui! attention au décalage d'indices en python) donc lancez:

In [5]:
# initialisation
M = [ 6*[0] for i in range(10)]
# remplissage
for i in range(10):
    for j in range(6):
        M[i][j] = (i+1)*(j+1)
        
for i in range(10):
    print(M[i])

[1, 2, 3, 4, 5, 6]
[2, 4, 6, 8, 10, 12]
[3, 6, 9, 12, 15, 18]
[4, 8, 12, 16, 20, 24]
[5, 10, 15, 20, 25, 30]
[6, 12, 18, 24, 30, 36]


*(bonus hp)*: l'écriture en compréhension permet de définir M en une seule ligne; contemplez:

In [ ]:
M = [[(i+1)*(j+1) for j in range(6)] for i in range(10)]

**exercice 2 - C'est-à-vous**: Construire une matrice M de taille 10×15 telle que la case (i,j) vaut:
$$M_{i,j} = (i-1)+(j-1)$$
Puis l'afficher. Attention, avec les décalages d'indice en python, on a : `M[i][j] = i + j`

ainsi sur la 1ere ligne (indice python i=0), on aura les nombres 0,1,2,…,14

puis sur la 2eme ligne (indice python i=1), on aura les nombres 1,2,3,…,15 etc.

---
**exemple 3** vous disposez d'une liste `L` qui contient 60 éléments (par exemple les 60 premiers nombres premiers). Vous voulez présenter la liste en une matrice `M` de 5 lignes par 12 colonnes; comment faire le remplissage de `M`?

In [10]:
from sympy import nextprime
L = [2]
for i in range(59):
    L.append(nextprime(L[-1]))
# les 20 premiers éléments de L
print(L[:20])

[2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37, 41, 43, 47, 53, 59, 61, 67, 71]


*réponse*: les numéros de colonnes vont de 0 à 11… ce sont donc les **restes** possibles modulo 12. 
Observons l'emplacement de l'élement d'indice 14:

$$M = \begin{pmatrix}
e_0& e_1 & e_2 & \cdots & e_{11}\\
e_{12} & e_{13}& e_{14} &\cdots & e_{23}\\
etc. & & & &
\end{pmatrix}$$

Cet élément est en $(2;3)$; si on numérote les lignes et colonnes à partir de 0, cela correspond à $L_1; C_2$.

Or $14 = \mathbf{1} \times 12 + \mathbf{2}$; on constate que les coordonnées de $e_{14}$ sont exactement `( quotient;reste)` de 14 par 12. 

Pour remplir `M`, on fait:

    Pour i de 0 à 9:
        Pour j de 0 à 11:
            M[i][j] = L[i*12 +j ]

In [11]:
# initialisation taille 5×12, que des 0
M = [ 12*[0] for i in range(5)]
# remplissage
for i in range(5):
    for j in range(12):
        M[i][j] = L[i*12 +j]
        
for i in range(5):
    print(M[i])

[2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37]
[41, 43, 47, 53, 59, 61, 67, 71, 73, 79, 83, 89]
[97, 101, 103, 107, 109, 113, 127, 131, 137, 139, 149, 151]
[157, 163, 167, 173, 179, 181, 191, 193, 197, 199, 211, 223]
[227, 229, 233, 239, 241, 251, 257, 263, 269, 271, 277, 281]


In [ ]:
# bonus: écriture dans l'autre sens
for k in range(120):
    i = k//12
    j = k%12
    M[i][j] = L[k]

**exercice 3 - C'est-à-vous**: on a encore $60 = 6\times 10$. En adaptant l'exemple précédent, construisez une matrice `N` de taille 10×6 contenant les 60 premiers nombres premiers (et affichez `N`)

***
### remplissage général en fonction de i et j

Dans l'exemple 3, nous avons pu identifié une *fonction* à 2 variables (i, j) permettant de calculer quel doit être le contenu de la case `M[i][j]`: 

    M[i][j] = L[i*12 +j]  <-- il fallait récupérer cet élément

<div class="alert alert-info">
    Étant donnée une fonction <code>f(i: int, j: int)</code>, l'écriture générale du remplissage de la matrice sera donc de la forme
    <pre><code>
    Pour i allant de 0 à (nombre de lignes - 1):
        Pour j allant de 0 à (nombre de colonnes -1):
            M[i][j] ← f(i, j)
    </code></pre>
→ bien retenir ce principe de double-boucle imbriquée.
</div>
***

**exercice 4**: soit `s` la variable suivante

In [12]:
# début du poème l'Albatros (Charles Baudelaire)
s = """Souvent, pour s’amuser, les hommes d’équipage
Prennent des albatros, vastes oiseaux des mers,
Qui suivent, indolents compagnons de voyage,
Le navire glissant sur les gouffres amers.
"""
print(len(s))

182


La taille de `s` se décompose en $182 = 2\times 7\times 13 = 14\times 13$ caractères (y compris les espaces et retours à la ligne).

Construisez une matrice `N` de taille 14×13 contenant chaque caractère de `s` dans ses cases et affichez-la.

***
(suite d'activité réservée aux plus rapides et téméraires)

*pour aller plus loin*: Soit `t` la variable suivante

In [15]:
t = "bienvenue chez les ch'tis est un fameux film de 2008"
print(len(t), len(t)//10, len(t)%10)

52 5 2


On constate qu'avec 52 caractères, cela ferait 5 lignes de 10 caractères pleines et il resterait 2 caractères seuls

Complétez la chaîne `t` avec suffisamment de caractères `*` pour une ligne de plus, puis construisez une matrice `O` de taille 6×10 contenant tous les caractères. (affichez `O` ensuite)

## Pour aller plus loin

* sur une liste `L`, la méthode `L.pop()` renvoie le dernier élément de `L` et l'enlève de la liste (même principe en javascript sur `Array`)

In [16]:
L = [3, 4, 5]
a = L.pop()
print(a, L)

5 [3, 4]


* sur une liste `L`, la méthode `L.insert(i, a)` insert l'élément `a` avant l'indice `i`

In [17]:
L.insert(1, a)
print(L)

[3, 5, 4]


**exercice**: avec `L = list("BTS SIO")`, créez une matrice contenant toutes les permutations circulaires des lettres: le haut de la matrice contiendra

    BTS SIO
    OBTS SI
    IOBTS S
    etc.